src/mod-double-xp-weekend.cpp

```diff
@@ -10,122 +10,252 @@ enum WeekendXP
+    //   on login.
+    
+    DoubleXpWeekend() { }
+    
+    // NOTE We need to access the DoubleXpWeekend logic from other
+    // places, so keep all the logic accessible via a singleton here,
+    // and have the `CommandScript` and the `PlayeScript` access this
+    // for the functionality they need.
+    static DoubleXpWeekend* instance()
     {
-        static ChatCommandTable commandTable =
-        {
-            { "weekendxp rate", HandleSetXPBonusCommand, SEC_PLAYER, Console::No },
-        };
-
-        return commandTable;
+        static DoubleXpWeekend instance;
+        return &instance;
     }
 
-    static bool HandleSetXPBonusCommand(ChatHandler* handler, int8 rate)
+    uint32 OnGiveXP(Player* player, uint32 originalAmount, uint8 xpSource) const
     {
-        Player* player = handler->GetPlayer();
-
-        int8 maxRate = sConfigMgr->GetOption<int8>("XPWeekend.MaxAllowedRate", 2);
+        if (!IsEventActive())
+        {
+            return originalAmount;
+        }
 
-        if (rate <= 0 || rate > maxRate)
+        if (ConfigQuestOnly() && xpSource != PlayerXPSource::XPSOURCE_QUEST && xpSource != PlayerXPSource::XPSOURCE_QUEST_DF)
         {
-            handler->PSendSysMessage(LANG_CMD_WEEKEND_XP_ERROR, maxRate);
-            handler->SetSentErrorMessage(true);
-            return true;
+            return originalAmount;
         }
 
-        player->UpdatePlayerSetting("mod-double-xp-weekend", SETTING_WEEKEND_XP_RATE, rate);
-        handler->PSendSysMessage(LANG_CMD_WEEKEND_XP_SET, rate);
+        if (player->GetLevel() >= ConfigMaxLevel())
+        {
+            return originalAmount;
+        }
 
-        return true;
+        float newAmount = (float)originalAmount * GetExperienceRate(player);
+        return (uint32) newAmount;
     }
-};
-
-class DoubleXpWeekend : public PlayerScript
-{
-public:
-    DoubleXpWeekend() : PlayerScript("DoubleXpWeekend") { }
 
-    void OnLogin(Player* player) override
+    void OnLogin(Player* player, ChatHandler* handler) const
     {
-        if (sConfigMgr->GetOption<bool>("XPWeekend.Announce", false))
+        // TODO I am assuming that this is always called when a character logs in...
+        // if that is not the case, thing migh get weird... Adding some asserts or warnings would be nice
+        // but I'm not sure how to handle "This shouldn't be happening but it is" kind of scenarios in acore
+        MigratePlayerSettings(player, handler);
+
+        if (ConfigAnnounce())
         {
-            if (IsEventActive() && !sConfigMgr->GetOption<bool>("XPWeekend.AlwaysEnabled", false))
+            if (IsEventActive() && !ConfigAlwaysEnabled())
             {
-                ChatHandler(player->GetSession()).PSendSysMessage("It's the weekend! Your XP rate has been set to: {}", GetExperienceRate(player));
+                float rate = GetExperienceRate(player);
+                // TODO not sure why some strings were defined in SQL scripts and others (like these)
```

Comment:
Yes, all of them should be in the database, however support for module strings is limited and it gets complicated once there are many modules and you dont want IDs to conflict